### PR TITLE
Extra markdown instruction editor / image uploading

### PR DESCRIPTION
**Pull request by bcjordan**
# Add ability to specify markdown instructions for levels

Previews markdown as changes happen.

markdown

On blockly side, gets rendered on the client.

short-long
# Add ability to upload images to an https://*.code.org location via editor (similar to GitHub)

screenshot-test

### CODE REVIEW

File: apps/src/templates/instructions.html.ejs [hunk 1]
```diff
@@ -1,9 +1,9 @@
-<% var msg = require('../locale'); %>
-
-<p class='dialog-title'><%= msg.puzzleTitle(locals) %></p>
-<% if (locals.instructions) { -%>
+<p class='dialog-title'><%= locals.puzzleTitle %></p>
+<% if (locals.renderedMarkdown) { -%>
+  <%- locals.renderedMarkdown %>
```
Comment (laurelfan) [inline apps/src/templates/instructions.html.ejs]: should this be <%= ?

File: dashboard/app/controllers/level_assets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  before_filter :authenticate_user!
+
+  def upload
+    filename = AWS::S3.upload_to_bucket('images.code.org', params[:file].original_filename, open(params[:file]), access: :public_read)
+    render json: { newAssetUrl: "https://images.code.org/#{filename}" }
+  end
+
+  private
+
+  def can_upload?
```
Comment (laurelfan) [inline dashboard/app/controllers/level_assets_controller.rb]: should probably check if they are an admin also

File: apps/src/templates/instructions.html.ejs (hunk 1 above)
Comment (bcjordan) [inline apps/src/templates/instructions.html.ejs]: [Oddly enough](http://stackoverflow.com/questions/10326950/render-a-variable-as-html-in-ejs/10330401#10330401), <%- is how to render something un-escaped

File: dashboard/app/controllers/level_assets_controller.rb (hunk 2 above)
Comment (bcjordan) [inline dashboard/app/controllers/level_assets_controller.rb]: that!

File: dashboard/app/views/levels/editors/_md_instructions.haml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+.field
+  = f.label :markdown_instructions
+  %p When specified, these instructions will render in the instructions popup.
+  %p The regular instruction field above will still display as the instructions below the game area.
+  %p Note: since markdown instructions do not currently get translated, please do not use outside of middle school, high school and PD.
+  %p You can drag-and-drop or paste images in to this box, and they will be uploaded to a suitable location.
+  = f.text_area :markdown_instructions, placeholder: 'Markdown Instructions', rows: 4
+  #markdown-instructions-preview
+
+%script{src: '/blockly/js/marked/marked.js', type: 'text/javascript'}
```
Comment (laurelfan) [inline dashboard/app/views/levels/editors/_md_instructions.haml]: does this work with the blockly package?

Comment (bcjordan) [inline dashboard/app/views/levels/editors/_md_instructions.haml]: Does appear to, e.g., http://studio.code.org/blockly/js/droplet/droplet-full.js

**Comment by sfilman** [comment]
Excitement!!! Does this change include the scrolling/respecting fixed page height work in the dialog, too?

**Comment by bcjordan** [comment]
@sfilman not yet! Coming up next though :)

**Pull request merged**